design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+
+Thus the graphics implementations must diverge, either by doing so within MapLibre Native or depending on another toolkit to do the rendering entirely. We are proposing the former approach, with support for multiple rendering SDKs in the MapLibre Native toolkit itself.
+
+This is the MapLibre Rendering Modularization Design Plan and the biggest goal is to pave the way for Metal support on iOS. But while we’ve got the hood open, so to speak, we’d also like to fix a few other things. So let’s get to it!
+
+But before that, a quick note on terminology.
+- "toolkit" usually refers to MapLibre Native, but may also be used in context for another project at a similar level.
+- "SDK" or Software Development Kit usually refers to the rendering library, like OpenGL or Metal, at least in this document.
+- "API" refers to an Application Programming Interface.  In this context it refers to the classes and methods made available to a developer.
+
+## Goals
+
+It is useful to split our goals into three sections to articulate what this proposal aims to accomplish:
+
+- **North Stars:** These guide the overall direction of the proposal, and are the overarching _why_.
+
+-  **Library refactors:** These are concrete changes we are proposing to the library to get to our North Stars.
+
+- **Evaluation metrics:** These are specific criteria we will accomplish through this proposal.
+
+### <a name="north-stars">North Stars</a>
+
+1. This refactor will allow migration to Metal on iOS with less overhead than a migration without this explicit modularization. In other words, migration will be substantially simplified.
+2. It should be possible for a developer familiar with the platform to modify visual rendering more easily than before.
+
+
+### <a name="core">Core Functionality</a>
+
+1. Shader (program) code will be implemented in a modular fashion, such that replacement of any given shader can happen without impacting other shaders.
+2. Shader (program) code may no longer be compiled from the [webgl source](https://github.com/maplibre/maplibre-gl-js/tree/main/src/shaders), and should be organized within the maplibre-gl-native repo (as a compile time option).
+3. Layer rendering logic should be replaceable at the dev level.  All the necessary support should be exposed and a developer should be able to take over representation of any given layer.
+4. Developers should be able to associate specific styles with specific layer rendering.  The style sheet should be able to call out the type of representation it would like beyond the default.
+5. Developers should be able to add new visual representations fed by existing geometry types.  Such as data display driven by tile sets or animated markers.
```

Comment:
`animated markers` are not a think in maplibre right now. I think this should be removed from this document.

Reply:
These are examples of the kind of thing a develop should be able to add externally without changing the core toolkit.  We're not proposing to implement them in this PR.